aptos-move/move-examples/rewards_pool/sources/rewards_pool.move

```diff
@@ -0,0 +1,288 @@
+        epoch: u64,
+    ): (vector<Object<Metadata>>, vector<u64>) acquires RewardsPool {
+        assert!(epoch < epoch::now(), EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH);
+        let all_rewards_tokens = reward_tokens(rewards_pool);
+        let non_empty_reward_tokens = vector[];
+        let reward_per_tokens = vector[];
+        let rewards_pool_data = safe_rewards_pool_data(&rewards_pool);
+        vector::for_each(all_rewards_tokens, |reward_token| {
+            let reward = rewards(claimer, rewards_pool_data, reward_token, epoch);
+            if (reward > 0) {
+                vector::push_back(&mut non_empty_reward_tokens, reward_token);
+                vector::push_back(&mut reward_per_tokens, reward);
+            };
+        });
+        (non_empty_reward_tokens, reward_per_tokens)
+    }
+
+    /// Allow a claimer to claim the rewards for a past epoch.
+    /// This returns a vector of rewards for all reward tokens.
+    public entry fun claim_rewards_entry(
+        claimer: &signer,
+        rewards_pool: Object<RewardsPool>,
+        epoch: u64,
+    ) acquires RewardsPool {
+        let rewards = claim_rewards(claimer, rewards_pool, epoch);
+        let claimer_addr = signer::address_of(claimer);
+        vector::for_each_reverse(rewards, |r| primary_fungible_store::deposit(claimer_addr, r));
+    }
+
+    /// Allow a claimer to claim the rewards for all tokens for a past epoch.
+    /// This returns a vector of rewards for all reward tokens in the same order as the rewards tokens.
+    /// If there's no reward for a specific reward token, the corresponding returned reward asset will be of zero
+    /// amount (created via fungible_asset::zero).
+    public fun claim_rewards(
+        claimer: &signer,
+        rewards_pool: Object<RewardsPool>,
+        epoch: u64,
+    ): vector<FungibleAsset> acquires RewardsPool {
+        assert!(epoch < epoch::now(), EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH);
+        let reward_tokens = reward_tokens(rewards_pool);
+        let rewards = vector[];
+        let claimer_addr = signer::address_of(claimer);
+        let rewards_data = unchecked_mut_rewards_pool_data(&rewards_pool);
+        vector::for_each(reward_tokens, |reward_token| {
+            let reward = rewards(claimer_addr, rewards_data, reward_token, epoch);
+            let reward_store = simple_map::borrow(&rewards_data.reward_stores, &reward_token);
+            if (reward == 0) {
+                vector::push_back(
+                    &mut rewards,
+                    fungible_asset::zero(fungible_asset::store_metadata(reward_store.store)),
+                );
+            } else {
+                // Withdraw the reward from the corresponding store.
+                let store_signer = &object::generate_signer_for_extending(&reward_store.store_extend_ref);
+                vector::push_back(&mut rewards, fungible_asset::withdraw(store_signer, reward_store.store, reward));
```

Comment:
Now I wonder maybe we should allow withdraw 0.

Reply:
May be we can introduce a withdraw_safe function. I still think disallowing withdrawing 0 is cleaner. There's usually much confusion when a function call is successful but doesn't do/return anything